src/nnet3/nnet-simple-component.cc

```diff
@@ -1956,6 +1957,652 @@ void NaturalGradientPerElementScaleComponent::Update(
+  const ConvolutionComponent *other =
+      dynamic_cast<const ConvolutionComponent*>(&other_in);
+  KALDI_ASSERT(other != NULL);
+  filter_params_.AddMat(alpha, other->filter_params_);
+  bias_params_.AddVec(alpha, other->bias_params_);
+}
+
+/*
+ This function transforms a vector of lists into a list of vectors,
+ padded with -1.
+ @param[in] The input vector of lists. Let in.size() be D, and let
+            the longest list length (i.e. the max of in[i].size()) be L.
+ @param[out] The output list of vectors. The length of the list will
+            be L, each vector-dimension will be D (i.e. out[i].size() == D),
+            and if in[i] == j, then for some k we will have that
+            out[k][j] = i. The output vectors are padded with -1
+            where necessary if not all the input lists have the same side.
+*/
+void RearrangeIndexes(const std::vector<std::vector<int32> > &in,
+                                                std::vector<std::vector<int32> > *out) {
+  int32 D = in.size();
+  int32 L = 0;
+  for (int32 i = 0; i < D; i++)
+    if (in[i].size() > L)
+      L = in[i].size();
+  out->resize(L);
+  for (int32 i = 0; i < L; i++)
+    (*out)[i].resize(D, -1);
+  for (int32 i = 0; i < D; i++) {
+    for (int32 j = 0; j < in[i].size(); j++) {
+      (*out)[j][i] = in[i][j];
+    }
+  }
+}
+
+void ConvolutionComponent::InderivPatchesToInderiv(
+    const CuMatrix<BaseFloat>& in_deriv_patches,
+    CuMatrixBase<BaseFloat> *in_deriv) const {
+
+  int32 num_x_steps = (1 + (input_x_dim_ - filt_x_dim_) / filt_x_step_);
+  int32 num_y_steps = (1 + (input_y_dim_ - filt_y_dim_) / filt_y_step_);
+  int32 num_frames = (*in_deriv).NumRows();
+  int32 filter_dim = filter_params_.NumCols();
+
+  // Compute the reverse column_map
+  std::vector<std::vector<int32> > reverse_column_map(in_deriv->NumCols());
+  for (int32 x_step = 0; x_step < num_x_steps; x_step++) {
+    for (int32 y_step = 0; y_step < num_y_steps; y_step++)  {
+      int32 patch_number = x_step * num_y_steps + y_step;
+      int32 patch_start_index = patch_number * filter_dim;
+      for (int32 x = 0, index = patch_start_index; x < filt_x_dim_; x++)  {
+        for (int32 y = 0; y < filt_y_dim_; y++)  {
+          for (int32 z = 0; z < input_z_dim_; z++, index++)  {
+            int32 vector_index;
+            if (input_vectorization_ == kZyx)  {
+              vector_index = ZyxVectorIndex(x_step * filt_x_step_ + x,
+                                            y_step * filt_y_step_ + y, z,
+                                            input_x_dim_, input_y_dim_,
+                                            input_z_dim_);
+            } else if (input_vectorization_ == kYzx)  {
+              vector_index = YzxVectorIndex(x_step * filt_x_step_ + x,
+                                            y_step * filt_y_step_ + y, z,
+                                            input_x_dim_, input_y_dim_,
+                                            input_z_dim_);
+            }
```

Comment:
The compiler can discover repeated expressions like that; what concerns me
more here is the repeated use of class variables like filt_x_step_, since
the compiler can't always know that they are unchanged- especially if other
functions are called in between- and it may repeatedly re-load them. It
would be better to have a local copy like
int32 filt_x_step = filt_x_step_;
Dan


On Tue, Oct 27, 2015 at 10:59 PM, Xingyu Na <notifications@github.com>
wrote:

> In src/nnet3/nnet-simple-component.cc
> <https://github.com/kaldi-asr/kaldi/pull/297#discussion_r43214191>:
>
> > +      int32 patch_start_index = patch_number * filter_dim;
> > +      for (int32 x = 0, index = patch_start_index; x < filt_x_dim_; x++)  {
> > +        for (int32 y = 0; y < filt_y_dim_; y++)  {
> > +          for (int32 z = 0; z < input_z_dim_; z++, index++)  {
> > +            int32 vector_index;
> > +            if (input_vectorization_ == kZyx)  {
> > +              vector_index = ZyxVectorIndex(x_step * filt_x_step_ + x,
> > +                                            y_step * filt_y_step_ + y, z,
> > +                                            input_x_dim_, input_y_dim_,
> > +                                            input_z_dim_);
> > +            } else if (input_vectorization_ == kYzx)  {
> > +              vector_index = YzxVectorIndex(x_step * filt_x_step_ + x,
> > +                                            y_step * filt_y_step_ + y, z,
> > +                                            input_x_dim_, input_y_dim_,
> > +                                            input_z_dim_);
> > +            }
>
> x_step * filt_x_step_ and y_step * filt_y_step_ can be reused.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/297/files#r43214191>.
>